bal_tools/subgraph.py

```diff
@@ -96,3 +102,136 @@ def get_first_block_after_utc_timestamp(self, timestamp: int) -> int:
+                and entry["chain"].lower() == chain.value.lower()
+                for item in entry["prices"]
+                if end_date_ts >= int(item["timestamp"]) >= start_date_ts
+            ]
+            if not prices:
+                return None
+            return TWAPResult(address=address, twap=sum(prices) / len(prices))
+
+        results = [calc_twap(addr) for addr in addresses]
+        return results[0] if len(results) == 1 else results
+
+    def get_twap_price_bpt(
+        self, pool_id: str, chain: GqlChain, date_range: DateRange
+    ) -> Decimal:
+        """
+        fetches the TWAP price of a pool's BPT over the given date range.
+
+        params:
+        - pool_id: the id of the pool
+        - chain: the chain network from GqlChain enum
+        - date_range: tuple of (start_date_ts, end_date_ts)
+
+        returns:
+        - Decimal(twap)
+        """
+        params = {
+            "chain": chain.value,
+            "id": pool_id,
+        }
+
+        token_data = self.fetch_graphql_data(
+            "core",
+            "get_pool_tokens",
+            params,
+            url="https://api-v3.balancer.fi",
+        )["poolGetPool"]
+
+        bpt_supply = Decimal(token_data["dynamicData"]["totalShares"])
+        token_addresses = [token["address"] for token in token_data["poolTokens"]]
+
+        twap_results = self.get_twap_price_token(
+            addresses=token_addresses,
+            chain=chain,
+            date_range=date_range,
+        )
+
+        pool_value = Decimal(0)
+        for token, twap_result in zip(token_data["poolTokens"], twap_results):
+            balance = Decimal(token["balance"])
+            pool_value += twap_result["twap"] * balance
+
+        return pool_value / bpt_supply
+
+    def calculate_aura_vebal_share(self, web3: Web3, block_number: int) -> Decimal:
+        """
+        Function that calculate veBAL share of AURA auraBAL from the total supply of veBAL
+        """
+        ve_bal_contract = web3.eth.contract(
+            address=web3.to_checksum_address(
+                "0xC128a9954e6c874eA3d62ce62B468bA073093F25"
+            ),
+            abi=get_abi("ERC20"),
+        )
+        total_supply = ve_bal_contract.functions.totalSupply().call(
+            block_identifier=block_number
+        )
+        aura_vebal_balance = ve_bal_contract.functions.balanceOf(
+            "0xaF52695E1bB01A16D33D7194C28C42b10e0Dbec2"  # veBAL aura holder
+        ).call(block_identifier=block_number)
+        return Decimal(aura_vebal_balance) / Decimal(total_supply)
+
+    def fetch_all_pools_info(self) -> List[Pool]:
+        """
+        Fetches all pools info from balancer graphql api
+        """
+        res = self.fetch_graphql_data(
+            "core", "vebal_get_voting_list", url="https://api-v3.balancer.fi"
+        )
+        return [Pool(**pool) for pool in res["veBalGetVotingList"]]
+
+    def get_balancer_pool_snapshots(
```

Comment:
still using the traditional endpoints for this, see: https://github.com/BalancerMaxis/bal_tools/issues/4#issuecomment-2113277322